### PR TITLE
chore(CSharp): add minima maxima

### PR DESCRIPTION
**Pull request by mheidari988**
**PR Checklist:**
- [x] My submission is formatted according to the guidelines in the [contributing guide](https://github.com/MakeContributions/DSA#contribution-guidelines)
- [x] My addition is on refer on the language README.md file
- [x] My addition does not have a spelling problem
- [x] My submission has a proper and user-friendly description of the algorithm
- [x] My submission has the time complexity of the algorithm
- [x] My submission has sample input-output of the program (NOT FOR PYTHON)

**What kind of change does this PR introduce?** (check at least one)

- [x] New algorithm

**Briefly describe the changes in this PR**

Minima Maxima is a searching algorithm to find local and global minima and maxima in a list of numbers

**Other information:**

### CODE REVIEW

File: algorithms/CSharp/src/Search/minima-maxima.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
```
Comment (ming-tsai) [inline algorithms/CSharp/src/Search/minima-maxima.cs:4]: Remove unused `using`

File: algorithms/CSharp/src/Search/minima-maxima.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            };
+
+            var lmins = GetLocalMinima(numbers);
+            var lmaxs = GetLocalMaxima(numbers);
+
+            foreach (var n in lmins)
+                Console.WriteLine($"Local Minima: {n}");
+
+            foreach (var n in lmaxs)
+                Console.WriteLine($"Local Maxima: {n}");
```
Comment (ming-tsai) [inline algorithms/CSharp/src/Search/minima-maxima.cs:24]: Add missing curly brackets
```suggestion
            foreach (var n in lmins)
            {
                Console.WriteLine($"Local Minima: {n}");
            }

            foreach (var n in lmaxs)
            {
                Console.WriteLine($"Local Maxima: {n}");
            }
```

File: algorithms/CSharp/src/Search/minima-maxima.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                    if (numbers[i - 1] <= numbers[i] && numbers[i] >= numbers[i + 1])
+                        result.Add(numbers[i]);
+
+                //Check last elements
+                if (numbers[^1] > numbers[^2])
+                    result.Add(numbers[^1]);
+            }
+
+            return result;
+        }
```
Comment (ming-tsai) [inline algorithms/CSharp/src/Search/minima-maxima.cs:79]: Add missing curly brackets
```suggestion
        public static List<int> GetLocalMinima(List<int> numbers)
        {
            var result = new List<int>();

            if (numbers.Count < 3)
            {
                result.Add(numbers.Min());
            }
            else
            {
                // Check first element
                if (numbers[0] < numbers[1])
                {
                    result.Add(numbers[0]);
                }

                //Loop middle elements
                for (int i = 1; i < numbers.Count - 1; i++)
                {
                    if (numbers[i - 1] >= numbers[i] && numbers[i] <= numbers[i + 1])
                    {
                        result.Add(numbers[i]);
                    }
                }

                //Check last elements
                if (numbers[^1] < numbers[^2])
                {
                    result.Add(numbers[^1]);
                }
            }

            return result;
        }

        public static List<int> GetLocalMaxima(List<int> numbers)
        {
            var result = new List<int>();

            if (numbers.Count < 3)
            {
                result.Add(numbers.Max());
            }
            else
            {
                // Check first element
                if (numbers[0] > numbers[1])
                {
                    result.Add(numbers[0]);
                }
                    
                //Loop middle elements
                for (int i = 1; i < numbers.Count - 1; i++)
                {
                    if (numbers[i - 1] <= numbers[i] && numbers[i] >= numbers[i + 1])
                    {
                        result.Add(numbers[i]);
                    }
                }

                //Check last elements
                if (numbers[^1] > numbers[^2])
                {
                    result.Add(numbers[^1]);
                }
            }

            return result;
        }
```

File: algorithms/CSharp/test/Search/minima-maxima.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+using System.Linq;
+
+namespace Algorithms.Tests.Search
+{
+    [TestFixture]
+    public class MinimaMaximaTest
+    {
+
+        [TestCase(new int[] { 3, 1, 2, 5, 6, 7, 4, 6, 9, 10 }, new int[] { 1, 4 })]
+        public void MinimaMaxima_GetListOfMinimas(int[] input, int[] result)
```
Comment (ming-tsai) [inline algorithms/CSharp/test/Search/minima-maxima.cs:12]: Rename the method name
```suggestion
        public void PassIntegerList_ShouldGetExpectedMinimas(int[] input, int[] result)
```

File: algorithms/CSharp/test/Search/minima-maxima.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            var expected1 = MinimaMaxima.GetLocalMinima(input.ToList());
+            Assert.AreEqual(expected1, result);
+        }
+
+        [TestCase(new int[] { 3, 1, 2, 5, 6, 7, 4, 6, 9, 10 }, new int[] { 3, 7, 10 })]
+        public void MinimaMaxima_GetListOfMaximas(int[] input, int[] result)
```
Comment (ming-tsai) [inline algorithms/CSharp/test/Search/minima-maxima.cs:19]: Rename the method name
```suggestion
        public void PassIntegerList_ShouldGetExpectedMaximas(int[] input, int[] result)
```

File: algorithms/CSharp/src/Search/minima-maxima.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+                //Check last elements
+                if (numbers[^1] < numbers[^2])
+                    result.Add(numbers[^1]);
+            }
+
+            return result;
+        }
+
+        public static List<int> GetLocalMaxima(List<int> numbers)
```
Comment (ming-tsai) [inline algorithms/CSharp/src/Search/minima-maxima.cs:54]: Rename the method name
```suggestion
        public static List<int> FindMaximas(List<int> numbers)
```

File: algorithms/CSharp/src/Search/minima-maxima.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            var lmaxs = GetLocalMaxima(numbers);
+
+            foreach (var n in lmins)
+                Console.WriteLine($"Local Minima: {n}");
+
+            foreach (var n in lmaxs)
+                Console.WriteLine($"Local Maxima: {n}");
+        }
+
+        public static List<int> GetLocalMinima(List<int> numbers)
```
Comment (ming-tsai) [inline algorithms/CSharp/src/Search/minima-maxima.cs:27]: Rename the method
```suggestion
        public static List<int> FindMinimas(List<int> numbers)
```

**Review by ming-tsai — CHANGES_REQUESTED** [review]
Hi @mheidari988,
I had added some suggestions to this pull request.

**Comment by mheidari988** [comment]
Thanks

File: algorithms/CSharp/src/Search/minima-maxima.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+            foreach (var n in lmins)
+            {
+                Console.WriteLine($"Local Minima: {n}");
+            }
+
+            foreach (var n in lmaxs)
+            {
+                Console.WriteLine($"Local Maxima: {n}");
+            }
```
Comment (ming-tsai) [inline algorithms/CSharp/src/Search/minima-maxima.cs:27]: Please call the method with a new name
```suggestion
            var minimas= FindMinimas(numbers);
            var maximas= FindMaximas(numbers);

            foreach (var minima in minimas)
            {
                Console.WriteLine($"Local Minima: {minima}");
            }

            foreach (var maxima in maximas)
            {
                Console.WriteLine($"Local Maxima: {maxima}");
            }
```

**Review by ming-tsai — CHANGES_REQUESTED** [review]
Hi @mheidari988,
I forgot to comment on this suggestion.

File: algorithms/CSharp/src/Search/minima-maxima.cs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        public static void Main()
+        {
+            var numbers = new List<int>
+            {
+                3,1,2,5,6,7,4,6,9,10
+            };
+
+            var minimas= FindMinimas(numbers);
+            var maximas= FindMaximas(numbers);
```
Comment (ming-tsai) [inline algorithms/CSharp/src/Search/minima-maxima.cs:17]:
```suggestion
            var minimas = FindMinimas(numbers);
            var maximas = FindMaximas(numbers);
```

File: algorithms/CSharp/test/Search/minima-maxima.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        public void PassIntegerList_ShouldGetExpectedMinimas(int[] input, int[] result)
+        {
+            var expected1 = MinimaMaxima.GetLocalMinima(input.ToList());
+            Assert.AreEqual(expected1, result);
+        }
+
+        [TestCase(new int[] { 3, 1, 2, 5, 6, 7, 4, 6, 9, 10 }, new int[] { 3, 7, 10 })]
+        public void PassIntegerList_ShouldGetExpectedMaximas(int[] input, int[] result)
+        {
+            var expected1 = MinimaMaxima.GetLocalMaxima(input.ToList());
```
Comment (ming-tsai) [inline algorithms/CSharp/test/Search/minima-maxima.cs:21]:
```suggestion
            var expected1 = MinimaMaxima.FindMaximas(input.ToList());
```

**Review by ming-tsai — APPROVED** [review]

**Pull request merged**